### PR TITLE
chore: Adds deprecation warnings for `app` parameter when using `to_asgi_response` method of responses

### LINKED ISSUE #2217: Refactor: remove `app` parameter from `Response.to_asgi_response`
The parameter `app` of `Response.to_asgi_response` is unused and redundant because it will always be the same as `request.app`.

It should be deprecated and marked for removal in the next major version.

---

## Funding
* If you would like to see an issue prioritized, make a pledge towards it!
* We receive the pledge once the issue is completed & verified

<a href="https://polar.sh/litestar-org/litestar/issues/2217">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/litestar-org/litestar/issues/2217/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by wevonosky**
### Pull Request Checklist

[//]: # "By submitting this issue, you agree to:"
[//]: # "- follow Litestar's [Code of Conduct](https://github.com/litestar-org/.github/blob/main/CODE_OF_CONDUCT.md)"
[//]: # "- follow Litestar's [Contribution Guidelines](https://litestar.dev/community/contribution-guide)"

- [X] New code has 100% test coverage
- [X] (If applicable) The prose documentation has been updated to reflect the changes introduced by this PR
- [X] (If applicable) The reference documentation has been updated to reflect the changes introduced by this PR

### Description
[//]: # "Please describe your pull request for new release changelog purposes"
[//]: # "Example: 'This pr adds the capability to use server-sent events, and documents it in the usage docs'"

This PR adds deprecation warnings for the use of `app` in the `to_asgi_response` methods of responses. The `app` parameter is redundant since the `request` parameter has the associated app available. 

-

### Close Issue(s)
[//]: # "Please add in issue numbers this pull request will close, if applicable"
[//]: # "Examples: Fixes #4321 or Closes #1234"

Closes #2217 
-

### CODE REVIEW

File: litestar/response/base.py [hunk 1]
```diff
@@ -401,6 +402,8 @@ def to_asgi_response(
             An ASGIResponse instance.
         """
 
+        warn_deprecation("2.1", "app", "parameter", removal_in="3.0.0")
```
Comment (provinzkraut) [inline litestar/response/base.py:405]: Please use keyword arguments here for everything. That makes it more readable! (=

Also, this warning should only be raised when the app parameter is actually used, so there's 3 things left to do here:

1. Change the parameter to be `app: Litestar | None`
2. Place the warning inside an `if app is not None` block
3. Remove the usage of the `app` parameter throughout the codebase

This applies to all the other instances as well.

**Review by provinzkraut — CHANGES_REQUESTED** [review]
Thanks for this! 

There are some changes required, which I detailed in a comment. Feel free to ask should you have any questions regarding the implementation of this!

File: litestar/response/base.py (hunk 1 above)
Comment (wevonosky) [reply litestar/response/base.py:405]: Ah thanks!

Comment (wevonosky) [reply litestar/response/base.py]: Ok I pushed up the changes and thanks for the guidance!

One thing to note is that [here](https://github.com/wevonosky/litestar/blob/main/litestar/handlers/http_handlers/_utils.py#L65) and [here](https://github.com/wevonosky/litestar/blob/main/litestar/handlers/http_handlers/_utils.py#L151) the app parameter is unused. I am realizing as I am typing this that the type into those two `handler` methods should probably also include `None`

Comment (wevonosky) [reply litestar/response/base.py]: Oof failing tests, working on it

**Comment by wevonosky** [comment]
@provinzkraut just pinging you for another review :)

File: litestar/response/base.py [hunk 2]
```diff
@@ -401,6 +402,15 @@ def to_asgi_response(
             An ASGIResponse instance.
         """
 
+        if app is not None:
+            warn_deprecation(
+                version="2.1",
+                deprecated_name="app",
+                kind="parameter",
+                removal_in="3.0.0",
+                info="app will be pulled from parameter 'request' instead",
```
Comment (provinzkraut) [inline litestar/response/base.py:411]:
```suggestion
                alternative="request.app",
```

File: litestar/response/file.py [hunk 3]
```diff
@@ -329,6 +330,15 @@ def to_asgi_response(
         Returns:
             A low-level ASGI file response.
         """
+        if app is not None:
+            warn_deprecation(
+                version="2.1",
+                deprecated_name="app",
+                kind="parameter",
+                removal_in="3.0.0",
+                info="app will be pulled from parameter 'request' instead",
```
Comment (provinzkraut) [inline litestar/response/file.py:339]:
```suggestion
                alternative="request.app",
```

File: litestar/response/redirect.py [hunk 4]
```diff
@@ -124,6 +125,15 @@ def to_asgi_response(
         cookies = self.cookies if cookies is None else filter_cookies(self.cookies, cookies)
         media_type = get_enum_string_value(self.media_type or media_type or MediaType.TEXT)
 
+        if app is not None:
+            warn_deprecation(
+                version="2.1",
+                deprecated_name="app",
+                kind="parameter",
+                removal_in="3.0.0",
+                info="app will be pulled from parameter 'request' instead",
```
Comment (provinzkraut) [inline litestar/response/redirect.py:134]:
```suggestion
                alternative="request.app",
```

File: litestar/response/streaming.py [hunk 5]
```diff
@@ -168,6 +169,14 @@ def to_asgi_response(
         Returns:
             An ASGIStreamingResponse instance.
         """
+        if app is not None:
+            warn_deprecation(
+                version="2.1",
+                deprecated_name="app",
+                kind="parameter",
+                removal_in="3.0.0",
+                info="app will be pulled from parameter 'request' instead",
```
Comment (provinzkraut) [inline litestar/response/streaming.py:178]:
```suggestion
                alternative="request.app",
```

File: litestar/response/template.py [hunk 6]
```diff
@@ -99,7 +100,16 @@ def to_asgi_response(
...
         type_encoders: TypeEncodersMap | None = None,
     ) -> ASGIResponse:
-        if not app.template_engine:
+        if app is not None:
+            warn_deprecation(
+                version="2.1",
+                deprecated_name="app",
+                kind="parameter",
+                removal_in="3.0.0",
+                info="app will be pulled from parameter 'request' instead",
```
Comment (provinzkraut) [inline litestar/response/template.py:109]:
```suggestion
                alternative="request.app",
```

File: tests/unit/test_response/test_response_to_asgi_response.py [hunk 7]
```diff
@@ -299,13 +299,21 @@ def handler() -> Template:
...
+    )
+
     with create_test_client(
         handler, template_config=TemplateConfig(engine=JinjaTemplateEngine, directory=tmp_path)
     ) as client:
         route: HTTPRoute = client.app.routes[0]
         route_handler = route.route_handlers[0]
         response = await route_handler.to_response(
-            data=route_handler.fn.value(), app=client.app, request=RequestFactory().get()
+            data=route_handler.fn.value(), app=client.app, request=RequestFactory(app=app).get()
```
Comment (provinzkraut) [inline tests/unit/test_response/test_response_to_asgi_response.py:316]: Why was this change necessary?

File: litestar/response/template.py [hunk 8]
```diff
@@ -99,7 +100,16 @@ def to_asgi_response(
...
+            )
+
+        if not request.app.template_engine:
```
Comment (provinzkraut) [inline litestar/response/template.py:112]: Maybe assign `app` to a separate variable here instead of accessing it via `request.app` each time.

**Review by provinzkraut — CHANGES_REQUESTED** [review]
Yup, this looks correct! A few minor things left to address (=

**Comment by Goldziher** [comment]
please resolve the last comment @wevonosky

File: tests/unit/test_response/test_response_to_asgi_response.py (hunk 7 above)
Comment (wevonosky) [reply tests/unit/test_response/test_response_to_asgi_response.py:316]: The `to_asgi_response`  in [this function](https://github.com/litestar-org/litestar/blob/44687e5b8aedb05002c79c46420fd5c00be7442c/litestar/response/template.py#L89) checks if the app template engine has been configured and raises an error otherwise. This test was raising that configuration error because the `RequestFactory` creates a request with a vanilla app that hasn't had a template engine configured. For that reason I am creating an app instance in this test and passing it through to `RequestFactory` so that the configuration error doesn't get raised and the test behaves like it did before.

**Comment by wevonosky** [comment]
@Goldziher and @provinzkraut I left a comment on why I had to modify that test. 

@all-contributors please add @wevonosky

**Comment by wevonosky** [comment]
@all-contributors please add @wevonosky for code

**Review by Goldziher — APPROVED** [review]
